File: 3748.count-stable-subarrays.java
#
# @lc app=leetcode id=3748 lang=java
#
# [3748] Count Stable Subarrays
#

# @lc code=start
class Solution {
    public long[] countStableSubarrays(int[] nums, int[][] queries) {
        int n = nums.length;
        int q = queries.length;
        long[] ans = new long[q];

        // Single pass: collect segments using lists
        java.util.List<Integer> segStartList = new java.util.ArrayList<>();
        java.util.List<Integer> segEndList = new java.util.ArrayList<>();
        for (int i = 0; i < n; ) {
            int j = i;
            while (j + 1 < n && nums[j + 1] >= nums[j]) ++j;
            segStartList.add(i);
            segEndList.add(j);
            i = j + 1;
        }
        int K = segStartList.size();
        int[] segStart = new int[K];
        int[] segEnd = new int[K];
        for (int k = 0; k < K; ++k) {
            segStart[k] = segStartList.get(k);
            segEnd[k] = segEndList.get(k);
        }

        // Precompute triangular numbers and prefix sums
        long[] segTotal = new long[K];
        long[] pref = new long[K + 1];
        for (int j = 0; j < K; ++j) {
            int len_ = segEnd[j] - segStart[j] + 1;
            segTotal[j] = (long) len_ * (len_ + 1) / 2;
            pref[j + 1] = pref[j] + segTotal[j];
        }

        // Process queries
        for (int qi = 0; qi < q; ++qi) {
            int L = queries[qi][0];
            int R = queries[qi][1];

            // Binary search for jL: largest j where segStart[j] <= L
            int low = 0, high = K - 1;
            while (low <= high) {
                int mid = low + (high - low) / 2;
                if (segStart[mid] <= L) {
                    low = mid + 1;
                } else {
                    high = mid - 1;
                }
            }
            int jL = high;

            // Binary search for jR: largest j where segStart[j] <= R
            low = 0;
            high = K - 1;
            while (low <= high) {
                int mid = low + (high - low) / 2;
                if (segStart[mid] <= R) {
                    low = mid + 1;
                } else {
                    high = mid - 1;
                }
            }
            int jR = high;

            long res;
            if (jL == jR) {
                int m = R - L + 1;
                res = (long) m * (m + 1) / 2;
            } else {
                // Left partial: L to segEnd[jL]
                int eL = segEnd[jL];
                long mL = (long) eL - L + 1;
                long numL = mL * (mL + 1) / 2;
                // Right partial: segStart[jR] to R
                int sR = segStart[jR];
                long mR = (long) R - sR + 1;
                long numR = mR * (mR + 1) / 2;
                // Middle full segments: jL+1 to jR-1
                long numMid = pref[jR] - pref[jL + 1];
                res = numL + numR + numMid;
            }
            ans[qi] = res;
        }

        return ans;
    }
}
# @lc code=end

File: 3748.count-stable-subarrays.go
#
# @lc app=leetcode id=3748 lang=golang
#
# [3748] Count Stable Subarrays
#

# @lc code=start
func countStableSubarrays(nums []int, queries [][]int) []int64 {
	n := len(nums)
	var runs [][2]int
	i := 0
	for i < n {
		j := i
		for j+1 < n && nums[j] <= nums[j+1] {
			j += 1
		}
		runs = append(runs, [2]int{i, j})
		i = j + 1
	}

	m := len(runs)
	belong := make([]int, n)
	for k := 0; k < m; k++ {
		for p := runs[k][0]; p <= runs[k][1]; p++ {
			belong[p] = k
		}
	}

	pref := make([]int64, m+1)
	for k := 0; k < m; k++ {
		s := runs[k][0]
		e := runs[k][1]
		ln := int64(e - s + 1)
		pref[k+1] = pref[k] + ln*(ln+1)/2
	}

	ans := make([]int64, len(queries))
	for qi, query := range queries {
		l := query[0]
		r := query[1]
		runL := belong[l]
		runR := belong[r]
		var total int64
		if runL == runR {
			ln := int64(r - l + 1)
			total = ln * (ln + 1) / 2
		} else {
			eL := runs[runL][1]
			lnL := int64(eL - l + 1)
			cl := lnL * (lnL + 1) / 2
			sR := runs[runR][0]
			lnR := int64(r - sR + 1)
			cr := lnR * (lnR + 1) / 2
			full := pref[runR] - pref[runL+1]
			total = cl + cr + full
		}
		ans[qi] = total
	}

	return ans
}
# @lc code=end